snowflake/objects/deploy.sql

```diff
@@ -0,0 +1,77 @@
+/*
+  This is the primary deployment script for RECOVER Snowflake objects.
+
+  Deployments are made to a specific environment. An environment is analogous
+  to a database. Account-level objects are not covered by this script. We assume
+  that the following objects have already been created at the account level:
+
+    - ROLE `RECOVER_DATA_ENGINEER`
+        * A ROLE with the appropriate privelages to create and deploy
+          objects in a database.
+    - WAREHOUSE `RECOVER_XSMALL`
+    - API INTEGRATION `RECOVER_GIT`
+        * An API INTEGRATION with an API_ALLOWED_PREFIXES value
+          containing 'https://github.com/Sage-Bionetworks/' and
+          API_PROVIDER = GIT_HTTPS_API.
+    - STORAGE INTEGRATION `RECOVER_PROD_S3`
+        * An S3 storage integration which allows access to the
+          S3 buckets in the RECOVER production account.
+
+  Additionally, we assume that the following databases have already been created
+  when deploying to the "staging" or "main" environment, respectively:
+
+    - DATABASE `RECOVER_STAGING`
+    - DATABASE `RECOVER_MAIN`
+
+  To deploy to an environment identified by your current git branch,
+  we can invoke this file with the Snowflake CLI like so:
+
+  `snow sql -D "environment=$(git rev-parse --abbrev-ref HEAD)" -f {this file}`
+*/
+
+/*
+  Configure the runner
+*/
+USE ROLE RECOVER_DATA_ENGINEER;
+USE WAREHOUSE RECOVER_XSMALL;
+
+/*
+  Configure and bootstrap the database if this is a dev environment.
+
+  Since we typically want to use the branch name as the environment identifier,
+  but only A-Za-z0-9 and _ and $ characters are allowed in Snowflake identifier
+  names and we have the convention of naming the branch after its associated
+  Jira ticket (e.g., etl-123), we substitute a likely hyphen in the `environment`
+  variable passed at runtime.
+*/
+SET safe_environment_identifier = (SELECT REPLACE('&{ environment }', '-', '_'));
+SET database_identifier = CONCAT('recover_', $safe_environment_identifier);
+
+EXECUTE IMMEDIATE
+$$
+BEGIN
+    IF ('&{ environment }' != 'main' AND '&{ environment }' != 'staging') THEN
+        -- This is a dev environment. Deploy from scratch.
+        CREATE OR REPLACE DATABASE IDENTIFIER($database_identifier);
+    END IF;
+END;
+$$;
+USE DATABASE IDENTIFIER($database_identifier);
+
+/*
+  Create an external stage over the RECOVER Git repository so that we can
+  use EXECUTE IMMEDIATE FROM statements.
+*/
+CREATE OR REPLACE GIT REPOSITORY recover_git_repository
```

Comment:
I think eventually, we will need ALTER this instead of create or replace in the future.

Reply:
@thomasyu888 Could you explain why we might need to maintain an existing `GIT REPOSITORY` object? By using `CREATE OR REPLACE` we guarantee that our repository is up to date without needing to worry about whether the repository has already been created – which would require a `FETCH` statement.

Reply:
@philerooski so, maybe im jumping ahead and just thinking of the `dev` and `prod` deployment.  I guess its not too much lift to create or replace these everytime we push to a branch.

I think I'm struggling a bit conceptually to see the production CI/CD pipeline so I'll wait for you to get the first data flowing in and we can revisit.

Reply:
Also I guess the issue is there's two parallel tracks here

1. Development workflow
2. Dev and prod deployment

What we are currently tackling here is (2), in that everytime you add a file (lets say Table) or add a scheduled task, the CI/CD is going to run through this entire stack to create the resource without creating all the other resources again. 

For (1), like you've mentioned, it would be CLONE DATABASE, then create all the new resources. 

Is that right?

Reply:
> I guess its not too much lift to create or replace these everytime we push to a branch.

No, it's not much overhead at all. And it's the only way to do stateless/declarative deployment in Snowflake – with the exception of objects which support the CREATE OR ALTER syntax.

> the CI/CD is going to run through this entire stack to create the resource without creating all the other resources again.

I'm a little confused about how you think this works, because you said above that we create or replace most objects every time we push to a branch, but this statement implies that we'll be able to know when a change is independent of already deployed objects, which would require a dependency graph.

> For (1), like you've mentioned, it would be CLONE DATABASE, then create all the new resource per a new branch. We do NOT load all the data up again for existing data., we just load data for the new resources or tables we're adding

That's right. Does it make things clearer if I put it like this: The deployment process for dev/staging/prod is exactly the same _except_ for the hard to replace objects, like tables with data.

Reply:
> we'll be able to know when a change is independent of already deployed objects, which would require a dependency graph.
Hmmm.  lets merge this and see how it plays out - _I think_ we are actually on the same page, but I think need to just _see it_ in action.

Snowflake takes care of what has already been deployed (hence the CREATE... IF NOT EXISTS, or CREATE OR ALTER statements)

Reply:
slight revision: The deployment process for dev/staging/prod is exactly the same except for the hard to replace objects **and their parent objects,** like tables with data, **and their schema and database.**